packages/protocol/docs/iprover.md

```diff
@@ -0,0 +1,60 @@
+# Implementing arbitrary token payments with IProver interface
+
+In this guide, we will outline the steps to implement a solution that enables prover pools to accept arbitrary ERC-20 tokens as payments for providing with proofs. This solution allows proposers to interact with pools (implementing `IProver`), agree on a price for proving a block, and make payments using ERC-20 tokens.
+
+NOTE: This works also with NFTs (ERC-721/ERC-1155) as well (applying the proper `approval`/`approvalForAll`), just because it might be less likely those will be used as payment methods, we highlighted the ERC-20. 
+
+## Prerequisites
+
+Before implementing this solution, make sure you have an existing ERC-20 token contract that you want to accept as payment in your ProverPool.
+
+## Implementation Steps
+
+### Step 1: An example ProverPool Contract
+
+Start by creating a ProverPool contract that implements the `IProver` interface. This interface should include the `onBlockAssigned` function, which will be called during `proposeBlock()`.
+A boilerplate (non-comprehensive) example:
+
+```solidity
+// Import necessary libraries and interfaces
+import "./IProver.sol";
+
+// Define the ProverPool contract
+contract ProverPool is IProver {
+
+    // ERC-20 address of the payment token
+    address ERC20TokenAddress;
+    
+    // Implement the onBlockAssigned function
+    function onBlockAssigned(
+        uint64 blockId,
+        TaikoData.BlockMetadataInput calldata input,
+        TaikoData.ProverAssignment calldata assignment
+    ) external {
+        // Decode the assignment data to retrieve signatures and other information
+        (bytes memory proverSignature, uint256 tokenAmount) = decodeAssignmentData(assignment);
+
+        // 1. Verify the prover signature is valid (off-chain verification)
+        require(isValidSignature(proverSignature, input), "Invalid prover signature");
+
+        // 2. Execute the transfer transaction
+        ERC20(ERC20TokenAddress).transferFrom(input.proposer, address(this), tokenAmount);
```

Comment:
```suggestion
        ERC20(ERC20TokenAddress).transferFrom(msg.sender, address(this), tokenAmount);
```
would suggest `msg.sender` to give out the ERC-20 tokens, `input.proposer` is the L2 account which will be corresponding L2 blocks's coinbase.

Reply:
Ok, i'd then add `tx.origin` bc. `msg.sender` is the TaikoL1.